### PR TITLE
Remove strict typeof byte/char dependencies in MemoryExtensions

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
This allows many of the vectorized code paths in `MemoryExtensions` (`IndexOf`, `Contains`, etc.) to be optimized for Ts other than byte and char. For instance, sbyte and ushort will now qualify for the performance optimizations.

The real reason for this is to prepare for the introduction of the `Utf8Char` primitive type, which we want to receive many of the same optimizations as `byte` receives in the vectorized code paths. With this change, there's now only one location where the `Utf8Char` check has to be added (all the way at the end of the file), rather than sprinkling additional `typeof(T) == typeof(Utf8Char)` checks all throughout this file.

Ideally we'd also be able to detect enums in here just for completeness.

WIP - unit tests have not been written for this.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
FYI, with this PR some methods like `SequenceCompareTo` still use direct `typeof(byte)` / `typeof(char)` comparisons. The reason for this is listed as a comment at the bottom of the file: the type check methods only say that the `Equals` method has a behavior consistent with byte / char; it doesn't say anything about `CompareTo` having such consistency.

File: src/System.Private.CoreLib/shared/System/MemoryExtensions.cs [hunk 1]
```diff
@@ -216,13 +216,13 @@ public static bool IsWhiteSpace(this ReadOnlySpan<char> span)
         public static bool Contains<T>(this Span<T> span, T value)
             where T : IEquatable<T>
         {
-            if (typeof(T) == typeof(byte))
+            if (IsTypeEquatableAsByte<T>())
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/MemoryExtensions.cs]: Would something like this work?
```csharp
if (default(T) != null && IsTypeEquatableAsByte<T>())
```

**Comment by GrabYourPitchforks** [comment]
@jkotas @AndyAyersMS What if we do the same trick we do for `RuntimeHelpers.IsOrContainsReferences<T>()`? Would implementing the `IsTypeEquatableAsByte<T>()` method in the VM solve the issue you both described?

**Comment by jkotas** [comment]
It would work, but it does not feel right to add this helper.

The question I ask for helper like these that support a regular framework code every library out there should be able to write: Is it something that we would expose as public API? I do not think that `IsTypeEquatableAsByte` helper would be ever exposed as a public API.

Note that these helpers have to be implemented up to 5 times, once for each runtime that uses the shared code (CoreCLR, CoreRT, ProjectN, Mono, Unity), so they are not "cheap".

**Comment by GrabYourPitchforks** [comment]
If we wanted to generalize it, we'd actually want something akin to C++'s [`std::is_trivial`](https://en.cppreference.com/w/cpp/types/is_trivial), but with the additional constraint that the type is `memcmp`-comparable. Keeping with the pattern that I'm terrible at choosing method names:

```cs
internal static bool IsMemcmpComparable<T>();
```

The behavior is that it'd return `true` only for primitives, enums, and structs of these that use the base implementation of `operator ==`. Then the caller could check `if (sizeof(T) == 1 && IsMemcmpComparable<T>()) { /* treat as BYTE */ }`. The issue with this is that I don't know how we'd plumb `Char8` / `Rune` / `IntPtr` / `UIntPtr` / etc. through here without special-casing them. Those types all implement `operator ==`, but they use a trivial equality comparison.

**Comment by jkotas** [comment]
Yes, something like that would be better.

**Comment by GrabYourPitchforks** [comment]
Is there a better method for determining whether a non-primitive type is "trivially comparable" other than special-casing it? I suppose I could introduce an `[IsTriviallyComparable]` internal attribute.

**Comment by jkotas** [comment]
I think we would just hardcode the logic in the runtime.

**Comment by GrabYourPitchforks** [comment]
@jkotas @AndyAyersMS I hard-coded the list of types and moved the logic into the VM. Also ran some quick sample apps with `MemoryExtensions.Contains<ushort>` and `MemoryExtensions.SequenceEqual<Char8>` to make sure they were going down their respective fast-path implementations. Is this along the lines of what you think a proper implementation would look like?

(I don't know if the particular changes I made to `MemoryExtensions` will put additional pressure on the JIT. I don't know how to measure that.)

File: src/vm/jitinterface.cpp [hunk 2]
```diff
@@ -7493,6 +7493,56 @@ bool getILIntrinsicImplementationForRuntimeHelpers(MethodDesc * ftn,
...
+            || methodTable == MscorlibBinder::GetClass(CLASS__CHAR)
+            || methodTable == MscorlibBinder::GetClass(CLASS__INT16)
+            || methodTable == MscorlibBinder::GetClass(CLASS__UINT16)
+            || methodTable == MscorlibBinder::GetClass(CLASS__INT32)
+            || methodTable == MscorlibBinder::GetClass(CLASS__UINT32)
+            || methodTable == MscorlibBinder::GetClass(CLASS__INT64)
+            || methodTable == MscorlibBinder::GetClass(CLASS__UINT64)
+            || methodTable == MscorlibBinder::GetClass(CLASS__INTPTR)
+            || methodTable == MscorlibBinder::GetClass(CLASS__UINTPTR)
+            || methodTable == MscorlibBinder::GetClass(CLASS__GUID)
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: Guid and Timespan are not "core" types. It feels odd to include them here. (The rest of the list is reasonable.)

File: src/vm/jitinterface.cpp [hunk 3]
```diff
@@ -7493,6 +7493,56 @@ bool getILIntrinsicImplementationForRuntimeHelpers(MethodDesc * ftn,
...
+        static const BYTE returnFalse[] = { CEE_LDC_I4_0, CEE_RET };
+
+        // Ideally we could detect automatically whether a type is trivially comparable
+        // (i.e., its operator == could be implemented via memcmp). But for now we'll
+        // do the simple thing and hardcode the list of types we know fulfill this contract.
+
+        if (methodTable == MscorlibBinder::GetClass(CLASS__BOOLEAN)
+            || methodTable == MscorlibBinder::GetClass(CLASS__BYTE)
+            || methodTable == MscorlibBinder::GetClass(CLASS__SBYTE)
+#ifdef FEATURE_UTF8STRING
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: It would be nice to have everything that has to do with Utf8String in the Utf8String PR.

**Comment by AndyAyersMS** [comment]
I don't see anything that would cause issues for the jit.

**Comment by jkotas** [comment]
cc @marek-safar New runtime dependency from runtime partition

**Comment by GrabYourPitchforks** [comment]
@jkotas @marek-safar Is there anything I can do here that would avoid the break on the Mono side of the house when this goes over? Perhaps move this `RuntimeHelpers` method into the shared partition and give it a default implementation?

**Comment by marek-safar** [comment]
@GrabYourPitchforks it's ok to have it as it's, although better name than quite confusing IsTriviallyComparable would be nice.

**Comment by jkotas** [comment]
@marek-safar Do you have suggestion for better name?

Some ideas: `IsTriviallyEquatable`, `EqualityComparesBytes`

**Comment by marek-safar** [comment]
The method does not compare anything so it should be something about the type, perhaps HasBlittableEquality or something like that

**Comment by GrabYourPitchforks** [comment]
Thanks for the suggestions! I'm going to go with `IsBitwiseEquatable` for now since the `IsXyz<T>()` pattern matches other methods on `RuntimeHelpers`, and IMO "blittable" seems a bit too marshal-related.

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test

**Pull request merged**